lls/arp.c

```diff
@@ -0,0 +1,215 @@
+}
+
+int
+process_arp(struct lls_cache *cache, struct gatekeeper_if *iface,
+	uint16_t tx_queue, struct rte_mbuf *buf,
+	struct ether_hdr *eth_hdr, struct arp_hdr *arp_hdr)
+{
+	struct in_addr target_ip;
+	int is_garp;
+
+	if (unlikely(arp_hdr->arp_hrd != rte_cpu_to_be_16(ARP_HRD_ETHER) ||
+		     arp_hdr->arp_pro != rte_cpu_to_be_16(ETHER_TYPE_IPv4) ||
+		     arp_hdr->arp_hln != ETHER_ADDR_LEN ||
+		     arp_hdr->arp_pln != sizeof(uint32_t)))
+		return -1;
+
+	/* TODO If sip is not in the same subnet as our IP address, drop. */
+
+	/*
+	 * Check if the target IP address is us or if it's
+	 * a Gratuitous ARP. If neither are true, then it
+	 * shouldn't have been sent to us.
+	 */
+	target_ip.s_addr = arp_hdr->arp_data.arp_tip;
+	is_garp = is_garp_pkt(arp_hdr);
+	if ((iface->ip4_addr.s_addr != target_ip.s_addr) && !is_garp)
+		return -1;
+
+	/* Update cache with source resolution, regardless of operation. */
+	lls_cache_add(cache, (void *)&arp_hdr->arp_data.arp_sip,
+		&arp_hdr->arp_data.arp_sha, iface->id);
+
+	if (is_garp)
+		return -1;
+
+	switch (rte_be_to_cpu_16(arp_hdr->arp_op)) {
+	case ARP_OP_REQUEST: {
+		uint16_t num_tx;
+
+		/* Set-up Ethernet header. */
+		ether_addr_copy(&eth_hdr->s_addr, &eth_hdr->d_addr);
+		ether_addr_copy(&iface->eth_addr, &eth_hdr->s_addr);
+
+		/* Set-up ARP header. */
+		arp_hdr->arp_op = rte_cpu_to_be_16(ARP_OP_REPLY);
+		ether_addr_copy(&arp_hdr->arp_data.arp_sha,
+			&arp_hdr->arp_data.arp_tha);
+		arp_hdr->arp_data.arp_tip = arp_hdr->arp_data.arp_sip;
+		ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_sha);
+		arp_hdr->arp_data.arp_sip = target_ip.s_addr;
+
+		/* Need to transmit reply. */
+		num_tx = rte_eth_tx_burst(iface->id, tx_queue, &buf, 1);
+		if (unlikely(num_tx != 1))
+			RTE_LOG(NOTICE, GATEKEEPER, "lls: ARP reply failed\n");
+		return 0;
+	}
+	case ARP_OP_REPLY:
+		/*
+		 * No further action required. Could check to make sure
+		 * arp_hdr->arp_data.arp_tha is equal to arp->ether_addr,
+		 * but there's nothing that can be done if it's wrong anyway.
+		 */
+		return -1;
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER, "lls: %s received an ARP packet with an unknown operation (%hu)\n",
+			__func__, rte_be_to_cpu_16(arp_hdr->arp_op));
+		return -1;
+	}
+}
+
+void
+print_arp_entry(struct lls_cache *cache, const void *ip_key,
+	const struct lls_cache_entry *entry)
+{
+	char ip_str[INET_ADDRSTRLEN];
+	struct in_addr ipv4_addr;
+
+	if (ip_key == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: trying to print an ARP entry with a NULL IP key\n");
+		return;
+	}
+
+	/* Keep IP address in network order for inet_ntop(). */
+	memcpy(&ipv4_addr, ip_key, cache->key_len);
+	if (inet_ntop(AF_INET, &ipv4_addr, ip_str, INET_ADDRSTRLEN) == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s: failed to convert a number to an IP address (%s)\n",
+			__func__, strerror(errno));
+		return;
+	}
+
+	if (entry == NULL)
+		RTE_LOG(INFO, GATEKEEPER,
+			"%s: unresolved (port %hhu) (stale=%s)\n",
+			ip_str, entry->port_id,
```

Comment:
OK -- the fix here is actually that we should actually test for `!entry->valid`, since an unresolved entry exists in memory (and has port and stale variables), it just doesn't have a valid Ethernet address.